### PR TITLE
fix(tests): pre-create the testing namespaces for cross-namespace resource references tests

### PR DESCRIPTION
**Pull request by michaelhtm**
Description of changes:
Making changes to [previous PR](https://github.com/aws-controllers-k8s/iam-controller/pull/112), to make sure namespaces are being tested

`@pytest.fixtures(scope="module")` don't pick up custom `@pytest.mark` so we changed the scope to `"function"`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: test/e2e/tests/test_references.py [hunk 1]
```diff
@@ -189,19 +191,22 @@ def test_role_policy_references(self, referring_role, referred_policy):
...
 
         time.sleep(1)
 
         policy_ref, policy_cr, policy_arn = referred_policy
 
+        time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
+        time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
+        time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
+        time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
+        time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
```
Comment (a-hilaly) [inline test/e2e/tests/test_references.py:209]: Do we need all these?

File: test/e2e/tests/test_references.py [hunk 2]
```diff
@@ -189,19 +191,22 @@ def test_role_policy_references(self, referring_role, referred_policy):
         role.wait_until_deleted(role_name)
     
     @pytest.mark.resource_data({'withNamespace': True})
-    def test_role_policy_namespace_references(self, referring_role, referred_policy):
+    def test_role_policy_namespace_references(self, referring_role, referred_policy, request):
 
         # create the resources in order that initially the reference resolution
         # fails and then when the referenced resource gets created, then all
         # resolutions eventually pass and resources get synced.
-        time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
```
Comment (a-hilaly) [inline test/e2e/tests/test_references.py:197]: are you sure we don't need this anymore?

File: test/e2e/tests/test_references.py [hunk 3]
```diff
@@ -95,26 +93,30 @@ def referring_role(request, referred_policy_name):
...
         data = marker.args[0]
-        if 'withNamespace' in data:
+        if 'withNamespace' in data and data['withNamespace']:
+            logging.info("ChangingNamespaceInPolicy")
+            k8s.create_k8s_namespace(
+                TESTING_NAMESPACE
+            )
+            time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
+            time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
+            time.sleep(CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS)
```
Comment (a-hilaly) [inline test/e2e/tests/test_references.py:115]: Do we need to sleep 3 times here?

File: test/e2e/tests/test_references.py [hunk 4]
```diff
@@ -95,26 +93,30 @@ def referring_role(request, referred_policy_name):
...
     filename = "policy_simple"
     namespace = "default"
 
+    replacements = REPLACEMENT_VALUES.copy()
+
     if marker is not None:
         data = marker.args[0]
-        if 'withNamespace' in data:
+        if 'withNamespace' in data and data['withNamespace']:
+            logging.info("ChangingNamespaceInPolicy")
```
Comment (a-hilaly) [inline test/e2e/tests/test_references.py:109]: logging not needed

File: test/e2e/tests/test_references.py [hunk 5]
```diff
@@ -95,26 +93,30 @@ def referring_role(request, referred_policy_name):
...
+            k8s.create_k8s_namespace(
+                TESTING_NAMESPACE
+            )
```
Comment (a-hilaly) [inline test/e2e/tests/test_references.py:112]: Do you want to create the namespace at this level? why not at the beginning of the test function?

File: test/e2e/tests/test_references.py [hunk 6]
```diff
@@ -34,31 +34,29 @@
 # Little longer to delete the policy since it's referred-to from the role...
 DELETE_POLICY_TIMEOUT_SECONDS = 30
 CHECK_WAIT_AFTER_REF_RESOLVE_SECONDS = 10
-TESTING_NAMESPACE = "custom_namespace"
+TESTING_NAMESPACE = random_suffix_name("policy-namespace", 24)
```
Comment (a-hilaly) [inline test/e2e/tests/test_references.py:37]: can we do something similar to L40?
```
@pytest.fixture(scope="module")
def referred_policy_name():
```

File: test/e2e/tests/test_references.py (hunk 1 above)
Comment (michaelhtm) [reply test/e2e/tests/test_references.py:209]: I will be testing with just one wait

File: .gitignore [hunk 7]
```diff
@@ -4,4 +4,5 @@
 .idea
 /docs/site
 bin
-build
\ No newline at end of file
+build
+/deploymentFiles
```
Comment (a-hilaly) [inline .gitignore:8]: are these specific to ACK

File: test/e2e/tests/test_references.py [hunk 8]
```diff
@@ -189,19 +191,22 @@ def test_role_policy_references(self, referring_role, referred_policy):
         role.wait_until_deleted(role_name)
     
     @pytest.mark.resource_data({'withNamespace': True})
-    def test_role_policy_namespace_references(self, referring_role, referred_policy):
+    def test_role_policy_namespace_references(self, referring_role, referred_policy, request):
```
Comment (a-hilaly) [inline test/e2e/tests/test_references.py:194]: is `request` needed in the parameters?

File: test/e2e/tests/test_references.py (hunk 5 above)
Comment (michaelhtm) [reply test/e2e/tests/test_references.py:112]: I want to only create the namespace if needed..I will get `namespace already exists` error if it's created for the second test.
Also referred_policy runs before the test does

File: .gitignore (hunk 7 above)
Comment (michaelhtm) [reply .gitignore:8]: no that's just a directory i created, will remove

File: test/e2e/tests/test_references.py (hunk 5 above)
Comment (a-hilaly) [reply test/e2e/tests/test_references.py]: fair enough, thank you!

**Review by a-hilaly — APPROVED** [review]
Sweet, thank you for the fixes!
/lgtm

**Pull request merged**